extensions/socket/internal.m

```diff
@@ -0,0 +1,657 @@
+    [skin logInfo:@"TCP socket connected"];
+    sock.userData = DEFAULT;
+
+    if (sock.connectCallback != LUA_NOREF) {
+        [skin pushLuaRef:refTable ref:sock.connectCallback];
+        sock.connectCallback = [skin luaUnref:refTable ref:sock.connectCallback];
+
+        if (![skin protectedCallAndTraceback:0 nresults:0]) {
+            const char *errorMsg = lua_tostring(skin.L, -1);
+            [skin logError:[NSString stringWithFormat:@"%s connect callback error: %s", USERDATA_TAG, errorMsg]];
+        }
+    }
+}
+
+- (void)socket:(HSAsyncSocket *)sock didAcceptNewSocket:(HSAsyncSocket *)newSocket {
+    [[LuaSkin shared] logInfo:@"Client socket connected"];
+
+    newSocket.userData = CLIENT;
+
+    @synchronized(self.connectedSockets) {
+        [self.connectedSockets addObject:newSocket];
+    }
+}
+
+- (void)socketDidDisconnect:(HSAsyncSocket *)sock withError:(NSError *)err {
+    if (sock.userData == CLIENT) {
+        [[LuaSkin shared] logInfo:@"Client disconnected"];
+
+        @synchronized(self.connectedSockets) {
+            [self.connectedSockets removeObject:sock];
+        }
+    } else if (sock.userData == SERVER) {
+        [[LuaSkin shared] logInfo:@"Server disconnected"];
+
+        @synchronized(self.connectedSockets) {
+            for (HSAsyncSocket *client in sock.connectedSockets){
+                [client disconnect];
+            }
+        }
+    } else [[LuaSkin shared] logInfo:@"Socket disconnected"];
+
+    sock.userData = nil;
+}
+
+- (void)socket:(HSAsyncSocket *)sock didWriteDataWithTag:(long)tag {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin logInfo:@"Data written to socket"];
+
+    if (self.writeCallback != LUA_NOREF) {
+        [skin pushLuaRef:refTable ref:self.writeCallback];
+        [skin pushNSObject: @(tag)];
+        self.writeCallback = [skin luaUnref:refTable ref:self.writeCallback];
+
+        if (![skin protectedCallAndTraceback:1 nresults:0]) {
+            const char *errorMsg = lua_tostring(skin.L, -1);
+            [skin logError:[NSString stringWithFormat:@"%s write callback error: %s", USERDATA_TAG, errorMsg]];
+        }
+    }
+}
+
+- (void)socket:(HSAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag  {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin logInfo:@"Data read from socket"];
+    NSString *utf8Data = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+
+    [skin pushLuaRef:refTable ref:self.readCallback];
+    [skin pushNSObject: utf8Data];
+    [skin pushNSObject: @(tag)];
+
+    if (![skin protectedCallAndTraceback:2 nresults:0]) {
+        const char *errorMsg = lua_tostring(skin.L, -1);
+        [skin logError:[NSString stringWithFormat:@"%s read callback error: %s", USERDATA_TAG, errorMsg]];
+    }
+}
+
+- (void)socket:(HSAsyncSocket *)sock didReceiveTrust:(SecTrustRef)trust completionHandler:(void (^)(BOOL))completionHandler {
+    // Allow TLS handshake without trust evaluation for self-signed certificates
```

Comment:
Added a comment there to clarify that this delegate method is only called when startTLS is invoked with` GCDAsyncSocketManuallyEvaluateTrust == YES`